lib/fio/replica.c

```diff
@@ -0,0 +1,761 @@
+		pthread_mutex_unlock(&mgmt_mtx);
+		return (1);
+	}
+	log_info("repl: replica server: %s:%d, volume name %s\n",
+	    mgmt_ack.ip, mgmt_ack.port, mgmt_ack.volname);
+
+	*port = mgmt_ack.port;
+	strncpy(host, mgmt_ack.ip, 256);
+
+	pthread_mutex_unlock(&mgmt_mtx);
+	return (0);
+}
+
+/*
+ * Opens and initializes data connection for zvol.
+ */
+static int fio_repl_open_file(struct thread_data *td, struct fio_file *f)
+{
+	struct repl_options *o = td->eo;
+	struct sockaddr_in addr;
+	short port;
+	char host[256] = "127.0.0.1";
+
+	port = o->port;
+	if (port && o->address)
+		strcpy(host, o->address);
+
+	// use mgmt connection to get host:port if not explicitly specified
+	if (!port) {
+		get_data_endpoint(td, f->file_name, &port, host);
+	}
+
+	f->fd = socket(AF_INET, SOCK_STREAM, 0);
+	if (f->fd < 0) {
+		td_verror(td, errno, "socket");
+		return (1);
+	}
+
+	if (o->nodelay) {
+		int optval = 1;
+
+		if (setsockopt(f->fd, IPPROTO_TCP, TCP_NODELAY,
+		    (void *) &optval, sizeof (int)) < 0) {
+			log_err("repl: cannot set TCP_NODELAY option on "
+			    "socket (%s), disable with 'nodelay=0'\n",
+			    strerror(errno));
+			return (1);
+		}
+	}
+
+	if (set_window_size(td, f->fd)) {
+		close(f->fd);
+		return (1);
+	}
+	if (set_mss(td, f->fd)) {
+		close(f->fd);
+		return (1);
+	}
+
+	memset(&addr, 0, sizeof (addr));
+	addr.sin_family = AF_INET;
+	addr.sin_addr.s_addr = htonl(INADDR_ANY);
+	addr.sin_port = htons(port);
+	if (inet_pton(AF_INET, host, &addr.sin_addr) <= 0) {
+		td_verror(td, errno, "inet_pton");
+		close(f->fd);
+		return (1);
+	}
+	log_info("repl: handshaking on data connection for volume %s\n",
+	    f->file_name);
+	if (connect(f->fd, (struct sockaddr *)&addr, sizeof (addr)) < 0) {
+		td_verror(td, errno, "connect");
+		close(f->fd);
+		return (1);
+	}
+
+	// send volume name we want to open to replica
+	if (write_handshake(td, f->fd, f->file_name) != 0) {
+		close(f->fd);
+		return (1);
+	}
+
+	return (0);
+}
+
+/*
+ * Close data connection.
+ */
+static int fio_repl_close_file(struct thread_data *td, struct fio_file *f)
+{
+	int rc;
+
+	rc = close(f->fd);
+	f->fd = -1;
+	return (rc);
+}
+
+static void fio_repl_terminate(struct thread_data *td)
+{
+	kill(td->pid, SIGTERM);
+}
+
+/*
+ * This is called before opening files and normally is used to retrieve file
+ * sizes. We use it for engine structure initialization and mgmt connection
+ * creation.
+ */
+static int fio_repl_setup(struct thread_data *td)
+{
+	struct repl_options *o = td->eo;
+	struct netio_data *nd;
+
+	if (!td->o.use_thread) {
+		log_err("repl: must set thread=1 when using repl plugin\n");
+		return (1);
+	}
+
+	nd = malloc(sizeof (*nd));
+	if (nd == NULL) {
+		log_err("repl: memory allocation failed\n");
+		return (1);
+	}
+	memset(nd, 0, sizeof (*nd));
+	nd->io_inprog = NULL;
+	nd->io_completed = calloc(td->o.iodepth, sizeof (struct io_u *));
+
+	// only create mgmt conn if it is needed
+	if (!o->port && mgmt_conn < 0) {
```

Comment:
lock required to check and set mgmt_conn

Reply:
fio takes special care to call setup hook sequentially so there can be no two threads running this code concurrently. For reference: https://github.com/axboe/fio/blob/62b858f646ad4fce971fbde21ff4859ec84bc281/backend.c#L2248

I will add comment to code to explain this as this could confuse people.